### ISSUE TITLE
Mixing defaults and functions

### ISSUE THREAD

**Issue by denismerigoux** [CatalaLang/catala#28]
Right now, the example `test/test_func/func.catala` is not passing. It fails on this assertion : https://github.com/CatalaLang/catala/blob/master/src/catala/lambda_calculus/lambda_interpreter.ml#L193.

This assertion is triggered because the default for the `f` wraps functions inside its justifications and consequences, as it's written here : https://github.com/CatalaLang/catala/blob/master/src/catala/catala_surface/desugaring.ml#L105. @nchataing you had set up things in a different manner but I changed them during the rewrite.

So two choices here : 

* either we keep the new desugaring style for default functions, but then we need to add a new rule in the lambda interpreter that first feeds the arguments to the functions inside the default and then evaluate it,
* either we push the function in front of the default, so then we evaluate the function first and then we get to the default. This would required changings the type of defs in our scope language because right now we assume all the definitions are default.

I'm done implementing for the week-end, so if somebody wants to pick up here please do :)

**Comment by denismerigoux** [comment]
Oh and the first thing to do with this function issue would be to switch the bindings to a locally nameless representation, instead of the `Uid.LocalVar.t` that we have now :)

**Comment by denismerigoux** [comment]
After consideration, we decided on the first solution of keeping functions inside the defaults justifications and consequences, with a special beta-reduction rule for default terms.

Indeed, it's more regular and generic formalization-wise. Moreover, it allows for better error handling because we can link the binder for each justification and consequence to its real position in the code. If we had merged the binder in front of the default, we would have had a problem when merging the positions.
